internal/security/secretstore/vmkencryption.go

```diff
@@ -0,0 +1,291 @@
+//
+// Copyright (c) 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+// ECCN: 5D002TSU
+//
+
+package secretstore
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/base64"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/edgexfoundry/edgex-go/internal/security/kdf"
+	"github.com/edgexfoundry/edgex-go/internal/security/pipedhexreader"
+	"github.com/edgexfoundry/edgex-go/internal/security/secretstoreclient"
+	"github.com/edgexfoundry/go-mod-secrets/pkg/token/fileioperformer"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+)
+
+/*
+
+THE FOLLOWING DISCLAIMER IS REQUIRED TO BE CARRIED WITH THE BELOW CODE
+DO NOT REMOVE EXCEPT BY PERMISSION OF THE AUTHOR
+
+Vault Master Key Encryption Feature
+
+The purpose of this feature is to provide a secure way to unlock Vault
+without requiring human intervention to supply the Vault Master Key.
+
+This feature requires a sufficiently secure source of randomness in order
+to unlock the Vault. This randomness should have at least 256 bits of entropy
+and be backed by hardware secure storage such as a TPM or TEE or a hardware
+security token.  If using a TPM, the secret should be bound to platform
+configuration registers to attest the system state.
+
+The randomness should be output as a string of hex-encoded octets to standard
+output and the path to the executable should be specified in the
+configuration.toml of security-secretstore-setup.
+
+In addition, the underlying platform that serves as the execution platform for
+EdgeX must be secured against the ability for an attacker to intercept the
+source of randomness between retrieving it out of hardware secure storage and
+passing it to security-secretstore-setup. This may entail use of secure boot,
+dm-verity encrypted partitions, volume encryption, orchestrator hardening,
+and other measures.
+
+*/
+
+const aesKeyLength = 32 // for AES-256
+
+type VMKEncryption struct {
+	fileOpener     fileioperformer.FileIoPerformer
+	pipedHexReader pipedhexreader.PipedHexReader
+	kdf            kdf.KeyDeriver
+	logger         logger.LoggingClient
+	encrypting     bool
+	ikm            []byte
+}
+
+// NewVMKEncryption - constructor
+func NewVMKEncryption(fileOpener fileioperformer.FileIoPerformer,
+	pipedHexReader pipedhexreader.PipedHexReader,
+	logger logger.LoggingClient,
+	kdf kdf.KeyDeriver) *VMKEncryption {
+
+	return &VMKEncryption{
+		fileOpener:     fileOpener,
+		pipedHexReader: pipedHexReader,
+		kdf:            kdf,
+		logger:         logger,
+		encrypting:     false,
+	}
+}
+
+// LoadIKM loads input key material from the specified path
+func (v *VMKEncryption) LoadIKM(ikmBinPath string) error {
+	if ikmBinPath == "" {
+		v.logger.Info("Not encrypting Vault master key")
+		return nil
+	}
+	ikm, err := v.pipedHexReader.ReadHexBytesFromExe(ikmBinPath, []string{})
+	if err != nil {
+		v.logger.Error(fmt.Sprintf("Error reading input key material from IKM_HOOK - encryption not enabled: %s", err.Error()))
```

Comment:
I don't think this module should be logging errors. It should add context to the error and return the error. Otherwise the error is getting logged again by the caller.